bpo-31699 Deadlocks in `concurrent.futures.ProcessPoolExecutor` with pickling error

tomMoral opened the pull request:
When using `concurrent.futures.ProcessPoolExecutor` with objects that are not picklable or unpicklable, several situations results in a deadlock, with the interpreter freezed.

This is the case for different scenario, for instance these [three scripts](https://gist.github.com/tomMoral/cc27a938d669edcf0286c57516942369). This PR propose to test the different failure scenario and fix the ones that results on deadlocks.

Overall, the goal is to make `concurrent.futures.ProcessPoolExecutor` more robust to faulty user code.

This work was done as part of the tommoral/loky#48 with the intent to re-use the executor in multiple independent part of the program, in collaboration with @ogrisel. See #1013 for more the details.



<!-- issue-number: bpo-31699 -->
https://bugs.python.org/issue31699
<!-- /issue-number -->

tomMoral committed:
TST no deadlocks with pickling crashes

- TST crash in CallItem unpickling
- TST crash in func call run
- TST crash in result pickling
the test include crashes with PythonError/SystemExist/SegFault

Also add more tests for race condition when a worker crashes

tomMoral committed:
ENH don't fail on result pickling errors

tomMoral committed:
DOC add NEWS entry

pitrou commented:
At a high level, I'm a bit surprised by the strategy adopted by this PR. Isn't it possible to keep a functioning executor in case an error occurs when (un)pickling? What does `multiprocessing.Pool` do (I don't remember)?

The crashing process problem is separate IMHO (and arguably less likely), though it does deserve to be handled.

tomMoral commented:
For unpickling errors, you have currently no way to get the `work_id` of the failure once the error occurred, that is why we resorted to flagging the executor as broken. Another strategy which could solve this would be to change the `_ResultItem/_CallItem` objects to have safe unpickle, which is guaranteed to give back at least the `work_id`, so in case of failures, we can just add the exception to this job. But I am not sure how to do it.
Note that `multiprocessing.Pool` is also broken when a result fails at unpickle. (see [here](https://github.com/python/cpython/blob/master/Lib/multiprocessing/pool.py#L470) ).
If you think this is necessary for this PR, I can look into it.

For pickling errors, this is much simpler and I should add something for the result pickling error to make sure we do not fail the executor in this case. The pickling error for the call item is already handled that way.

tomMoral commented:
After thinking a bit, my main concern about safe pickling/unpickling of `_ResultItem/_CallItem` is with the fact that we are agnostic of the `Pickler` objects used in the `call_queue` and `results_queue`. This should be the `_ForkingPickler`, but this could be changed, depending on the context, as we do in the `loky` project. So I am not sure if it is the right solution.

ogrisel commented:
@tomMoral can you please summarize the change you did in the last commit. The tests deadlock now. I think this is no longer robust to the `ExitAtUnpickle` case (which I think is fine but it should be removed from the test suite).

tomMoral commented:
@ogrisel In the last commit, I just removed the crash detection thread and the cases in the test suite not handled anymore. It works locally but it seems to fail deterministically on travis with `clang`. I cannot reproduce it even with a docker image from travis.

ogrisel commented:
I managed to reproduce the deadlock (without -j4 and with --timeout=300):

```
test_crash (test.test_concurrent_futures.ProcessPoolForkserverExecutorDeadlockTest) ... error at task pickle
exit at task unpickle
error at task unpickle
crash at task unpickle
crash during func execution on worker
exit during func execution on worker
error during func execution on worker
crash during result pickle on worker
exit during result pickle on worker
error during result pickle on worker
error during result unpickle in result_handler

Timeout (0:05:00)!
Thread 0x00007fa7a0a88700 (most recent call first):
  File "/home/parietal/ogrisel/cpython/Lib/threading.py", line 296 in wait
  File "/home/parietal/ogrisel/cpython/Lib/multiprocessing/queues.py", line 224 in _feed
  File "/home/parietal/ogrisel/cpython/Lib/threading.py", line 865 in run
  File "/home/parietal/ogrisel/cpython/Lib/threading.py", line 917 in _bootstrap_inner
  File "/home/parietal/ogrisel/cpython/Lib/threading.py", line 885 in _bootstrap

Thread 0x00007fa7a24a1700 (most recent call first):
  File "/home/parietal/ogrisel/cpython/Lib/selectors.py", line 415 in select
  File "/home/parietal/ogrisel/cpython/Lib/multiprocessing/connection.py", line 920 in wait
  File "/home/parietal/ogrisel/cpython/Lib/concurrent/futures/process.py", line 308 in _queue_management_worker
  File "/home/parietal/ogrisel/cpython/Lib/threading.py", line 865 in run
  File "/home/parietal/ogrisel/cpython/Lib/threading.py", line 917 in _bootstrap_inner
  File "/home/parietal/ogrisel/cpython/Lib/threading.py", line 885 in _bootstrap

Thread 0x00007fa7a748b700 (most recent call first):
  File "/home/parietal/ogrisel/cpython/Lib/multiprocessing/synchronize.py", line 96 in __enter__
  File "/home/parietal/ogrisel/cpython/Lib/multiprocessing/queues.py", line 357 in put
  File "/home/parietal/ogrisel/cpython/Lib/concurrent/futures/process.py", line 567 in shutdown
  File "/home/parietal/ogrisel/cpython/Lib/test/test_concurrent_futures.py", line 873 in test_crash
  File "/home/parietal/ogrisel/cpython/Lib/unittest/case.py", line 615 in run
  File "/home/parietal/ogrisel/cpython/Lib/unittest/case.py", line 663 in __call__
  File "/home/parietal/ogrisel/cpython/Lib/unittest/suite.py", line 122 in run
  File "/home/parietal/ogrisel/cpython/Lib/unittest/suite.py", line 84 in __call__
  File "/home/parietal/ogrisel/cpython/Lib/unittest/suite.py", line 122 in run
  File "/home/parietal/ogrisel/cpython/Lib/unittest/suite.py", line 84 in __call__
  File "/home/parietal/ogrisel/cpython/Lib/unittest/suite.py", line 122 in run
  File "/home/parietal/ogrisel/cpython/Lib/unittest/suite.py", line 84 in __call__
  File "/home/parietal/ogrisel/cpython/Lib/unittest/runner.py", line 176 in run
  File "/home/parietal/ogrisel/cpython/Lib/test/support/__init__.py", line 1892 in _run_suite
  File "/home/parietal/ogrisel/cpython/Lib/test/support/__init__.py", line 1936 in run_unittest
  File "/home/parietal/ogrisel/cpython/Lib/test/test_concurrent_futures.py", line 1185 in test_main
  File "/home/parietal/ogrisel/cpython/Lib/test/support/__init__.py", line 2068 in decorator
  File "/home/parietal/ogrisel/cpython/Lib/test/libregrtest/runtest.py", line 176 in runtest_inner
  File "/home/parietal/ogrisel/cpython/Lib/test/libregrtest/runtest.py", line 140 in runtest
  File "/home/parietal/ogrisel/cpython/Lib/test/libregrtest/main.py", line 379 in run_tests_sequential
  File "/home/parietal/ogrisel/cpython/Lib/test/libregrtest/main.py", line 458 in run_tests
  File "/home/parietal/ogrisel/cpython/Lib/test/libregrtest/main.py", line 536 in _main
  File "/home/parietal/ogrisel/cpython/Lib/test/libregrtest/main.py", line 510 in main
  File "/home/parietal/ogrisel/cpython/Lib/test/libregrtest/main.py", line 585 in main
  File "/home/parietal/ogrisel/cpython/Lib/test/__main__.py", line 2 in <module>
  File "/home/parietal/ogrisel/cpython/Lib/runpy.py", line 85 in _run_code
  File "/home/parietal/ogrisel/cpython/Lib/runpy.py", line 193 in _run_module_as_main
```

ogrisel commented:
Here is another run, this time the GC of the executor triggers the deadlock (via the wearkref callback):

```
test_crash (test.test_concurrent_futures.ProcessPoolForkserverExecutorDeadlockTest) ... error at task pickle
exit at task unpickle
error at task unpickle
crash at task unpickle
crash during func execution on worker
exit during func execution on worker
error during func execution on worker
crash during result pickle on worker
exit during result pickle on worker
error during result pickle on worker
error during result unpickle in result_handler
Timeout (0:01:00)!
Thread 0x00007f8bf9956700 (most recent call first):
  File "/home/parietal/ogrisel/cpython/Lib/multiprocessing/synchronize.py", line 96 in __enter__
  File "/home/parietal/ogrisel/cpython/Lib/multiprocessing/queues.py", line 357 in put
  File "/home/parietal/ogrisel/cpython/Lib/concurrent/futures/process.py", line 486 in weakref_cb
  File "/home/parietal/ogrisel/cpython/Lib/unittest/case.py", line 643 in run
  File "/home/parietal/ogrisel/cpython/Lib/unittest/case.py", line 663 in __call__
  File "/home/parietal/ogrisel/cpython/Lib/unittest/suite.py", line 122 in run
  File "/home/parietal/ogrisel/cpython/Lib/unittest/suite.py", line 84 in __call__
  File "/home/parietal/ogrisel/cpython/Lib/unittest/suite.py", line 122 in run
  File "/home/parietal/ogrisel/cpython/Lib/unittest/suite.py", line 84 in __call__
  File "/home/parietal/ogrisel/cpython/Lib/unittest/suite.py", line 122 in run
  File "/home/parietal/ogrisel/cpython/Lib/unittest/suite.py", line 84 in __call__
  File "/home/parietal/ogrisel/cpython/Lib/unittest/runner.py", line 176 in run
  File "/home/parietal/ogrisel/cpython/Lib/test/support/__init__.py", line 1892 in _run_suite
  File "/home/parietal/ogrisel/cpython/Lib/test/support/__init__.py", line 1936 in run_unittest
  File "/home/parietal/ogrisel/cpython/Lib/test/test_concurrent_futures.py", line 1186 in test_main
  File "/home/parietal/ogrisel/cpython/Lib/test/support/__init__.py", line 2068 in decorator
  File "/home/parietal/ogrisel/cpython/Lib/test/libregrtest/runtest.py", line 176 in runtest_inner
  File "/home/parietal/ogrisel/cpython/Lib/test/libregrtest/runtest.py", line 140 in runtest
  File "/home/parietal/ogrisel/cpython/Lib/test/libregrtest/main.py", line 379 in run_tests_sequential
  File "/home/parietal/ogrisel/cpython/Lib/test/libregrtest/main.py", line 458 in run_tests
  File "/home/parietal/ogrisel/cpython/Lib/test/libregrtest/main.py", line 536 in _main
  File "/home/parietal/ogrisel/cpython/Lib/test/libregrtest/main.py", line 510 in main
  File "/home/parietal/ogrisel/cpython/Lib/test/libregrtest/main.py", line 585 in main
  File "/home/parietal/ogrisel/cpython/Lib/test/__main__.py", line 2 in <module>
  File "/home/parietal/ogrisel/cpython/Lib/runpy.py", line 85 in _run_code
  File "/home/parietal/ogrisel/cpython/Lib/runpy.py", line 193 in _run_module_as_main
```

tomMoral committed:
ENH independent com channel for wakeup to avoid deadlocks in shutdown

ogrisel commented:
Note: I also had to use `taskset -c 0 python -m test ...` to trigger the race condition. The problem is the lock of the result handler queue that has not yet been released when the worker process is terminated.

ogrisel commented:
@tomMoral this PR is now conflicting with master.

tomMoral commented:
This PR has been rebased on `master` and I think it is ready for review.

ogrisel commented on Lib/concurrent/futures/process.py:
```diff
@@ -75,12 +76,36 @@
 _global_shutdown = False
 
+
+# This constants control the maximal wakeup. If a job is submitted to the
```
Typo: "This constant control*s*"

ogrisel commented on Lib/concurrent/futures/process.py:
```diff
@@ -261,30 +319,57 @@ def shutting_down():
-        if reader in ready:
-            result_item = reader.recv()
+        # Wait for a result to be ready in the result_queue while checking
+        # that worker process are still running.
```
typo: "that all worker process*es* are still running"

ogrisel commented on Lib/concurrent/futures/process.py:
```diff
@@ -424,34 +516,48 @@ def __init__(self, max_workers=None, mp_context=None,
+        # result_queue state. This avoid deadlocks caused by the non
+        # transmission of wakeup signal when a worker died with the
+        # _result_queue write lock.
+        self._wakeup = _Sentinel()
```
I think this should better be renamed to `_queue_management_thread_sentinel` to be more explicit.

ogrisel commented on Lib/concurrent/futures/process.py:
```diff
@@ -461,10 +567,12 @@ def weakref_cb(_, q=self._result_queue):
             self._queue_management_thread.daemon = True
             self._queue_management_thread.start()
-            _threads_queues[self._queue_management_thread] = self._result_queue
+            _threads_queues[self._queue_management_thread] = self._wakeup
```
`_threads_queues` should be renamed to `_threads_sentinels`.

ogrisel commented on Lib/test/test_concurrent_futures.py:
```diff
@@ -759,6 +763,236 @@ def test_ressources_gced_in_workers(self):
+            p.terminate()
+        executor.shutdown(wait=True)
+        print(f"\nTraceback:\n {tb}", file=sys.__stderr__)
+        self.fail(f"Deadlock executor:\n\n{tb}")
```
Phrasing: "Executor deadlock:"

ogrisel commented on Lib/test/test_concurrent_futures.py:
```diff
@@ -759,6 +763,236 @@ def test_ressources_gced_in_workers(self):
+            from signal import SIGTERM as SIGKILL
+        try:
+            os.kill(pid, SIGKILL)
+            time.sleep(.01)
```
Why is this sleep necessary? Shouldn't it be part of the caller instead? Or alternatively, kill several times until the OS replies that the process is dead:

```python
n_trials = 3

for i in range(n_trials):
    try:
        os.kill(pid, SIGKILL)
    except (ProcessLookupError, PermissionError):
        break
    time.sleep(0.01)
else:
    raise RuntimeError(f"Could not kill process {pid} after {n_trials} trials")
```

tomMoral replied:
This sleep increases the chances for this process to be killed by another process and give time to make sure that the `BrokenProcessPool` error occurs before the end of the current function.
I think that if a process fail to kill another one, it means that it was already shutdown so we should not make multiple tries.

tomMoral committed:
CLN correct some typos

tomMoral committed:
CLN improve test_crash_races

ogrisel commented on Lib/concurrent/futures/process.py:
```diff
@@ -253,6 +309,8 @@ def _queue_management_worker(executor_reference,
         result_queue: A ctx.SimpleQueue of _ResultItems generated by the
             process workers.
+        wakeup: A _Sentinel to allow waking up the queue_manager_thread from
+            the main Thread and avoid deadlocks caused by broken queues.
```
I think we should rename `wakeup` to `thread_sentinel` everywhere in this PR but maybe @pitrou has another opinion?

ogrisel commented on Lib/concurrent/futures/process.py:
```diff
@@ -72,15 +73,39 @@
+_POLL_TIMEOUT = .001
+
+
+class _Sentinel:
```
In retrospect I find this class name confusing as it's very different from the worker process' `.sentinel` attribute. Maybe we should find a better name  e.g. `_ThreadNotifier`.

tomMoral replied:
Maybe putting `_ThreadWakeup` and `thread_wakeup` everywhere?

ogrisel replied:
Actually, it should be possible to rewrite this class to use an anonymous `os.pipe` to make it possible use it directly within the call to `multiprocessing.connection.wait` without relying on manual polling with `_POLL_TIMEOUT`.

That should make it possible to reduce the initial job submission latency and reduce CPU usage when idling on most platforms at the cost of only 2 additional fds.

I am not sure if waiting on a pipe is possible and efficient under Windows though.

ogrisel replied:
BTW, I am fine with the name `_ThreadWakeup` personally.

ogrisel replied:
Here is what I meant about using a "waitable" pipe:

```python
>>> import os
>>> from multiprocessing.connection import wait
>>> r, w = os.pipe()
>>> reader = os.fdopen(r, mode='rb')
>>> writer = os.fdopen(w, mode='wb')
>>> wait([reader], timeout=1)
[]  # empty list after waiting for 1 second
>>> writer.write(b'\x00')
1
>>> writer.flush()
>>> wait([reader])
[<_io.BufferedReader name=149>]   # no wait here: instant result
```

ogrisel replied:
@pitrou do you know if using a pipe for waking up the call to `wait` in the main loop of the queue management thread is guaranteed to work on all supported platforms?

pitrou replied:
I don't think so. You have to use a dedicated primitive on Windows.

pitrou replied:
What are you trying to do exactly?

tomMoral replied:
To avoid dealing with different primitive, we can use a `multiprocessing.Pipe`.  
But I think we ruled that out before because of performance issues (_eg_ submitting many small tasks).

@pitrou : the goal of the `wakeup` object is to separate the waking up of the `queue_management_thread` from the `result_queue` to ensure that we can still communicate with it if the queue is deadlocked. This object right now is a flag shared between the threads, and thus we cannot use it in the `wait` primitive with it.

ogrisel replied:
This is only required for communication between the main thread (or some other user-spawn thread) and the queue management thread. We don't really need interprocess communication primitives here but it would be nice if we can avoid doing the 1ms wait loop and only rely on `select` to find which event to process next: result in the queue, process died or executor terminated or new process to create because we are submitting a job and the process count need to be adjusted. The last 2 kinds of events need this waitable "wakeup" communication channel between threads.

pitrou replied:
Why don't you use a Queue or Event?

ogrisel replied:
`Event` is not waitable via `multiprocessing.connection.wait` (at line https://github.com/python/cpython/pull/3895/files#diff-d24fedf7a1cf058e9e4166d89f2bb378R356). I was reluctant to use a `multirocessing.queues.Queue` to avoid adding the overhead of an additional helper feeder thread. But actually, we could probably use a `SimpleQueue` whose `_reader` attribute should be waitable in a cross-platform manner (I think).

ogrisel replied:
Under the hood it's using a `multiprocessing.connection.Pipe` but the `SimpleQueue` API is higher level.

ogrisel replied:
Note that `SimpleQueue` is a `Pipe` protected by a `Lock`. To wakeup the queue management thread when submitting a new task or when shutting down the executor, we really need to just send an arbitrary marker such as a single byte and flush. In master we are sending a None object via the result queue at the moment. So there is no need for a lock. We could therefore slightly reduce the overhead in `submit` by directly using a `Pipe` instead of pickling `None` + lock acquire and release.

pitrou replied:
I meant a simple threading Queue, not a heavy-weight multiprocessing object.

A thought: how about a separate thread that monitors processes using their sentinels and then signals the main management thread using a threading Queue?

tomMoral replied:
The usage ot a `queue.Queue` does not really present advantages over the actual implementation which relies on the current `_Sentinel` object, as we only use it as a notifier, we do not care if there is multiple messages. I think that we fixed the `_POLL_TIMEOUT` such it has the same precision as the `wait` function so the overhead should be minim compared to the usage of a `Pipe` but it is a bit messier.
The benefit of using a `multiprocessing.Pipe` is that we can wait on it with `select`. This way, there is a single call responsible for the wait on the `_result_queue`, the sentinels and the `thread_wakeup`.

The experience of adding an extra thread to manage the processes has been done with `multiprocessing.Pool`. I do not remember perfectly why but it seems to me that it lead to extra complexity in the code. But I can look into it if you feel it is a better design for the executor.

ogrisel replied:
> I meant a simple threading Queue, not a heavy-weight multiprocessing object.

But then it's not possible to use a call to `wait` to wait at the same time for the next submit, for the next result or a worker process to die. The current solution in this PR is introducing a polling mechanism with `_POLL_TIMEOUT`. I am now thinking that we should rather not use that solution to avoid CPU usage when polling on an idling executor and to reduce latency when submitting new tasks.

> A thought: how about a separate thread that monitors processes using their sentinels and then signals the main management thread using a threading Queue?

I think the code can stay simpler (easier to debug and maintain) by not introducing yet a new thread. Also the queue management thread has to monitor the results in the `multiprocessing.queues.SimpleQueue`. We cannot wait on the union of a `multiprocessing.queues.SimpleQueue` and a `threading.Queue` via a single call to `multiprocessing.connection.wait` so that would not remove the need for `_POLL_TIMEOUT`.

pitrou replied:
I think we removed most (all?) sources of polling in multiprocessing some time ago. I don't want to reintroduce any; spurious wakeups can be very bad for power consumption on mobile devices (think laptops). I'll let you find the best solution to avoid them :-)

tomMoral committed:
ENH remove polling in queue_management_thread

- Use a  to communicate between MainThread and QueueManager
  as it is waitable and avoid polling in the thread.
- Add timeout on ExecutorTest to avoid deadlocks freezing the test suite.
- Rename the  object to .

ogrisel commented on Lib/concurrent/futures/process.py:
```diff
@@ -72,15 +73,30 @@
 # threads/processes finish.
 
-_threads_queues = weakref.WeakKeyDictionary()
+_threads_wakeup = weakref.WeakKeyDictionary()
```
I would use the plural: "_threads_wakeups".

ogrisel commented on Lib/concurrent/futures/process.py:
```diff
@@ -72,15 +73,30 @@
+    def __init__(self):
+        self._reader, self._writer = mp.Pipe(duplex=False)
+
+    def set(self):
```
Maybe this method could actually be called `wakeup()`. Although I am fine with `set` as well personally.

ogrisel commented on Lib/concurrent/futures/process.py:
```diff
@@ -132,6 +149,25 @@ def __init__(self, work_id, fn, args, kwargs):
+            # work_item can be None if another process terminated (see above)
+            if work_item is not None:
+                work_item.future.set_exception(e)
+                del work_item
```
I think there is no need to del a local variable just before the end of the method.

ogrisel commented on Lib/concurrent/futures/process.py:
```diff
@@ -261,30 +312,54 @@ def shutting_down():
+                result_item = result_reader.recv()
+                is_broken = False
+            except BaseException as e:
+                traceback.print_exc()
```
Instead of printing the exception in the thread via stderr, I think we could store as in a local variable named "cause" that is set to `None` otherwise.

And later:

```python

bpe = BrokenProcessPool(message)
if cause is not None:
    bpe.__cause__ = cause
for work_id, work_item in pending_work_items.items():
    work_item.future.set_exception(bpe)
    del work_item
```

ogrisel commented on Lib/concurrent/futures/process.py:
```diff
@@ -424,34 +507,49 @@ def __init__(self, max_workers=None, mp_context=None,
+        # Permits to wake_up the queue_manager_thread independently of
+        # result_queue state. This avoid deadlocks caused by the non
+        # transmission of wakeup signal when a worker died with the
+        # _result_queue write lock.
```
May I suggest the following comment instead:

_ThreadWakeup is a communication channel used to interrupt
the wait of the main loop of queue_manager_thread from another
thread (e.g. when calling executor.submit or executor.shutdown).
We do not use the _result_queue to send the wakeup signal to the 
queue_manager_thread as it could result in a deadlock if a worker process
dies with the _result_queue write lock still acquired.

ogrisel commented on Lib/concurrent/futures/process.py:
```diff
@@ -424,34 +507,49 @@ def __init__(self, max_workers=None, mp_context=None,
-        def weakref_cb(_, q=self._result_queue):
-            q.put(None)
         if self._queue_management_thread is None:
+            # When the executor gets lost, the weakref callback will wake up
```
"gets lost" => "gets garbage collected"

ogrisel commented on Lib/concurrent/futures/process.py:
```diff
@@ -424,34 +507,49 @@ def __init__(self, max_workers=None, mp_context=None,
-            q.put(None)
         if self._queue_management_thread is None:
+            # When the executor gets lost, the weakref callback will wake up
+            # the queue management thread.
```
so that it can terminate if there is no pending work item.

ogrisel reviewed (commented):
Thanks for the updated wake implementation @tomMoral . Here is another batch of comments.

tomMoral committed:
CLN fix some typo+better bpe msg+fix some test

ogrisel commented on Lib/concurrent/futures/process.py:
```diff
@@ -253,6 +300,9 @@ def _queue_management_worker(executor_reference,
             process workers.
+        thread_wakeup: A _ThreadWakeup to allow waking up the
+            queue_manager_thread from the main Thread and avoid deadlocks
+            caused by broken queues.
```
permanently locked queues.

ogrisel commented on Lib/concurrent/futures/process.py:
```diff
@@ -261,30 +311,56 @@ def shutting_down():
+        elif wakeup_reader in ready:
+            thread_wakeup.clear()
+            is_broken = False
+            result_item = None
```
I think we should always clear the `wakeup_reader` even if there is a `result_item` in the result queue:

```python
cause = None
is_broken = True
result_item = None

if wakeup_reader in ready:
    thread_wakeup.clear()
    is_broken = False

if result_reader in ready:
    try:
        result_item = result_reader.recv()
        is_broken = False
    except BaseException as e:
        cause = traceback.format_exception(type(e), e, e.__traceback__)
        is_broken = True

...
```

Otherwise might not empty the wakeup pipe quickly enough in case of a large number of submissions of very short tasks.

ogrisel commented on Lib/test/test_concurrent_futures.py:
```diff
@@ -114,8 +116,13 @@ def setUp(self):
     def tearDown(self):
+        self.timer.cancel()
+        self.timer.join()
+        del self.timer
```
This `del` is not really necessary I think.

tomMoral replied:
The `del` is necessary to make sure the thread is `gc`-ed before the end of the test method.
Without it, I get some `stopped` dangling threads with the python test suite.

ogrisel replied:
Calling `del` does not guarantee that the deleted object will be instantaneously garbage collected though.

> Without it, I get some stopped dangling threads with the python test suite.

This is reported as a warning by the test runner? If so, maybe add a comment to explain that the call to `del` is necessary to avoid the dangling threads detector of the test runner to issue a warning.

ogrisel reviewed (approved):
Some more minor comments but other than that I think it's ready for final review by @pitrou or another core dev.

tomMoral committed:
ENH faster thread_wakeup clear+typo

ogrisel commented on Lib/concurrent/futures/process.py:
```diff
@@ -357,9 +357,9 @@ def shutdown_worker():
-            thread_wakeup.clear()
             is_broken = False
             result_item = None
+        thread_wakeup.clear()
```
Indeed this is simpler :)

tomMoral commented:
There is an issue with the CI on travis which is not run: https://travis-ci.org/python/cpython/jobs/308902863
This appears to be due to invalid commit range. Not sure if I did something wrong with the PR? Or maybe I need to rebase?

tomMoral committed:
CLN comment on del self.timer

pitrou commented on Lib/concurrent/futures/process.py:
```diff
@@ -72,15 +73,30 @@
+    items = list(_threads_wakeups.items())
+    for t, thread_wakeup in items:
+        thread_wakeup.wakeup()
     for t, q in items:
```
Perhaps update variable name `q`.

pitrou commented on Lib/concurrent/futures/process.py:
```diff
@@ -132,6 +149,24 @@ def __init__(self, work_id, fn, args, kwargs):
+            tb = traceback.format_exception(type(e), e, e.__traceback__)
+            e.__cause__ = _RemoteTraceback('\n"""\n{}"""'.format(''.join(tb)))
+            work_item = self.pending_work_items.pop(obj.work_id, None)
+            # work_item can be None if another process terminated (see above)
```
I don't understand what "see above" refers to.

pitrou commented on Lib/concurrent/futures/process.py:
```diff
@@ -261,30 +311,56 @@ def shutting_down():
+        n_children_alive = sum(p.is_alive() for p in processes.values())
+        n_children_to_stop = n_children_alive
+        n_sentinels_sent = 0
+        # Sent the right number of sentinels, to make sure all children are
```
Typo: "send" not "sent".

pitrou commented on Lib/concurrent/futures/process.py:
```diff
@@ -424,34 +509,52 @@ def __init__(self, max_workers=None, mp_context=None,
         self._initializer = initializer
         self._initargs = initargs
 
+        # Management threads
```
Singular "thread", no?

pitrou commented on Lib/concurrent/futures/process.py:
```diff
@@ -424,34 +509,52 @@ def __init__(self, max_workers=None, mp_context=None,
+        # when calling executor.submit or executor.shutdown). We do not use the
+        # _result_queue to send the wakeup signal to the queue_manager_thread
+        # as it could result in a deadlock if a worker process dies with the
+        # _result_queue write lock still acquired.
```
Nice explanation :-)

pitrou commented on Lib/concurrent/futures/process.py:
```diff
@@ -261,30 +311,56 @@ def shutting_down():
-            result_item = reader.recv()
-        else:
+        # Wait for a result to be ready in the result_queue while checking
+        # that all worker processes are still running.
```
...or that we are shutting down, no?

pitrou commented on Lib/test/_test_multiprocessing.py:
```diff
@@ -1029,6 +1029,42 @@ def __reduce__(self):
+                    obj.on_queue_feeder_error_was_called = True
+
+        not_serializable_obj = NotSerializable()
+        with test.support.captured_stderr():
```
What is the captured_stderr() for? A warning? A log message?

pitrou commented on Lib/test/test_concurrent_futures.py:
```diff
@@ -134,6 +143,23 @@ def _prime_executor(self):
         for f in futures:
             f.result()
 
+    def _fail_on_deadlock(self, executor=None):
```
Is this useful? You could instead pass `--timeout` to `python -m test`.

tomMoral replied:
`--timeout` does not work with the `-j` option so it lead to some deadlocks in the CI.
This option is safer as it will always work.

pitrou replied:
I don't want to add test-specific code to workaround issues in our test runner. Perhaps @vstinner wants to take a look why `--timeout` doesn't work with `-j`, but I don't think this PR should be complicated by this.

tomMoral replied:
After some investigation:

- The `--timeout` does work with `-j` but the dumped tracebacks are non informative: they qre tb of the runner and it happens before the indiviual test subprocesses have a chance to timeout by themselves. We did change that in 65f3aba so the timeout is now per test only.

- However, this does not fix all the issues. The `faulthandler.dump_traceback_later` with `exit=True` hang when the exited test process also has subprocesses by itself. In our test, the subprocesses will never terminate in case of a deadlock, as the `queue_manager_thread` does not exist anymore at this point. The solution would be to make `faulthandler` recursively terminate all subprocesses when exiting. There is a related open issue for `faulthandler` : https://bugs.python.org/issue12413

We remove the test-specific timers but we kept the timeout for `test_crashes` in order to avoid stalling CI for 1hour+ with tests that try to trigger deadlocks.

pitrou commented on Lib/test/test_concurrent_futures.py:
```diff
@@ -759,6 +788,205 @@ def test_ressources_gced_in_workers(self):
+    """Induces a segfault"""
+    import faulthandler
+    faulthandler.disable()
+    faulthandler._sigsegv()
```
I'm not sure whether this is reliable or not. Is `os.abort()` sufficient instead?
Perhaps @vstinner can help.

tomMoral replied:
`os.abort` dump traceback on `stderr` by default. This would make the test report unnecessarily noisy, so I would rather keep the current implementation.

pitrou commented on Lib/test/test_concurrent_futures.py:
```diff
@@ -759,6 +788,205 @@ def test_ressources_gced_in_workers(self):
+        ]
+        for func, args, error, name in crash_cases:
+            with self.subTest(name):
+                # skip the test involving pickle errors with manager as it
```
I don't understand this comment, as I don't see any skip below.

pitrou commented on Lib/test/test_concurrent_futures.py:
```diff
@@ -759,6 +788,205 @@ def test_ressources_gced_in_workers(self):
+            from signal import SIGKILL
+        except ImportError:
+            from signal import SIGTERM as SIGKILL
+        # Try to kill a process in the pool, if it is not finished yet
```
The problem is, this may kill another process if the pid is reused...

pitrou commented on Lib/test/test_concurrent_futures.py:
```diff
@@ -759,6 +788,205 @@ def test_ressources_gced_in_workers(self):
+        # Give some time for the Executor to detect the failure
+        time.sleep(.5)
+
+    def test_crash_races(self):
```
Is it important to have this test in addition to `test_crash`?
If so, then I suggest you look at `test_killed_child` to see how to kill a child in a safe and cross-platform way.

tomMoral replied:
Indeed it is redundant. I removed it.

ogrisel commented:
When googling the "fatal: Invalid symmetric difference expression" error message on travis I found this post by @pitrou on the python committers mailing list:

https://mail.python.org/pipermail/python-committers/2017-July/004717.html

but no one replied :) Maybe try a rebase.

tomMoral committed:
CLN improved comments for ProcessPoolExecutor

tomMoral committed:
CLN remove specific per-test timer

tomMoral committed:
CLN remove unnecessary code and improve test_shutdown_deadlock

pitrou commented on Lib/test/libregrtest/runtest_mp.py:
```diff
@@ -149,8 +149,6 @@ def run(self):
 def run_tests_multiprocess(regrtest):
     output = queue.Queue()
     pending = MultiprocessIterator(regrtest.tests)
-    test_timeout = regrtest.ns.timeout
```
I don't think we want to remove this in this PR. If there's a problem with parallel tests and the `--timeout` option, better open a separate issue about it.

tomMoral replied:
I removed the commit from the PR and I will open another one for this change.

pitrou commented on Lib/test/test_concurrent_futures.py:
```diff
@@ -759,6 +766,175 @@ def test_ressources_gced_in_workers(self):
+            tb = f.read()
+        for p in executor._processes.values():
+            p.terminate()
+        executor.shutdown(wait=True)
```
If the executor is in a deadlock state, `shutdown(wait=True)` will never return, right?

tomMoral replied:
As we kill the processes just before, this should be safe. In this case, if the `queue_management_worker` is alive, it will flag the `ProcessPoolExecutor` as broken and clean up the state and if it is not, the call to shutdown won't be blocking and the ressources will still be freed.

pitrou commented on Lib/test/test_concurrent_futures.py:
```diff
@@ -759,6 +766,175 @@ def test_ressources_gced_in_workers(self):
+        import faulthandler
+        from tempfile import TemporaryFile
+        with TemporaryFile(mode="w+") as f:
+            faulthandler.dump_traceback(file=f)
```
Why not `file=sys.stderr`?

ogrisel replied:
If I recall correctly, `faulthandler` actually needs a system-level file handle, not just a Python-level file object.

tomMoral replied:
This part of the code is called with `test.support.captured_stderr`. In this case, the `sys.stderr` object does not have a fileno and `fault_handler.dump_traceback` fails. So we resorted to this to allow getting the traceback in the exception. Another option is to dump the traceback on `sys.stdout`.

pitrou replied:
I see. Do you think this whole code is still useful? I understand it was useful for you to debug the PR, but now that it seems to pass, is there a reason for keeping it?

tomMoral replied:
The reason to keep it would be to have an informative test report in case of a regression introducing a deadlock. As deadlocks potentially happen randomly, it is always good to have a traceback asap when it happens. So I would rather keep it.
But we could move it to `sys.stdout` if you think the usage of a tempfile introduces unnecessary complexity. The test report would still have the information but in a degraded format: not in the test report, but in the test stdout.

pitrou commented on Lib/test/test_concurrent_futures.py:
```diff
@@ -94,6 +95,7 @@ def tearDown(self):
 
 
 class ExecutorMixin:
+    timeout = 30
```
This doesn't seem used anymore.

pitrou commented on Lib/test/test_concurrent_futures.py:
```diff
@@ -116,6 +118,8 @@ def setUp(self):
         self._prime_executor()
 
     def tearDown(self):
+        # Remove the reference to self.timer to avoid the thread_cleanup
```
Also this comment doesn't seem necessary anymore. Or am I missing something?

pitrou commented on Lib/test/test_concurrent_futures.py:
```diff
@@ -369,12 +373,15 @@ def test_del_shutdown(self):
-        call_queue.close()
+        # Make sure that the queue management thread was properly finished
+        # and the queue was closed by the shutdown process
+        queue_management_thread.join()
```
It's already joined 6 lines above. The second `join()` call is a no-op, no?

pitrou commented on Lib/test/test_concurrent_futures.py:
```diff
@@ -759,6 +766,180 @@ def test_ressources_gced_in_workers(self):
 
+def hide_process_stderr():
+    import io
+    setattr(sys, "stderr", io.StringIO())
```
Is there a subtlety that I'm missing, or can this simply be written as `sys.stderr = io.StringIO()`?

pitrou commented on Lib/test/test_concurrent_futures.py:
```diff
@@ -759,6 +766,180 @@ def test_ressources_gced_in_workers(self):
+
+
+class ExitAtPickle(object):
+    """Bad object that triggers a segfault at pickling time."""
```
The docstring is wrong here.

pitrou commented on Lib/test/test_concurrent_futures.py:
```diff
@@ -759,6 +766,180 @@ def test_ressources_gced_in_workers(self):
+
+
+class ErrorAtPickle(object):
+    """Bad object that triggers a segfault at pickling time."""
```
And here as well.

pitrou commented on Lib/test/test_concurrent_futures.py:
```diff
@@ -759,6 +766,180 @@ def test_ressources_gced_in_workers(self):
+
+
+class ErrorAtUnpickle(object):
+    """Bad object that triggers a process exit at unpickling time."""
```
And here too :-)

tomMoral replied:
The classic copy/paste mistake... Thanks for pointing it out! :-)

pitrou commented on Lib/test/test_concurrent_futures.py:
```diff
@@ -759,6 +766,180 @@ def test_ressources_gced_in_workers(self):
+                    executor.shutdown(wait=True)
+
+    @classmethod
+    def _test_getpid(cls, a):
```
This doesn't seem used anymore?

pitrou commented on Lib/test/test_concurrent_futures.py:
```diff
@@ -759,6 +766,180 @@ def test_ressources_gced_in_workers(self):
+
+
+class ExecutorDeadlockTest:
+    # If ExecutorDeadlockTest takes more than 100secs to complete, it is very
```
"100secs" isn't in sync with the number below.

tomMoral replied:
The comment was a duplicate from the one 3 lines below so I removed it.

pitrou reviewed (changes requested):
@ogrisel thanks for pinging me (did you amend previous commits? I didn't see any recent commits). There are still a couple nits left.

tomMoral committed:
CLN test comments + remove unused code

pitrou committed:
Improvement NEWS wording
